src/libcollections/vec.rs

```diff
@@ -1748,13 +1741,74 @@ impl<T> Drop for IntoIter<T> {
     }
 }
 
-/// An iterator that drains a vector.
+/// A trait for draining a vector.
+///
+/// See the documentation of `Vec::drain`.
+pub trait DrainRange {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T>;
+}
+
+impl DrainRange for Range<usize> {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        assert!(self.start <= self.end, "Range not increasing");
+        assert!(self.end <= vec.len(), "Range out of bounds");
+        unsafe {
+            let tail = vec.len() - self.end;
+            vec.set_len(tail + self.start);
+            let ptr = vec.as_mut_ptr();
+            let (start, end) = if mem::size_of::<T>() == 0 {
+                // make sure that start is not null
+                ((self.start + 1) as *mut T, (self.end + 1) as *mut T)
+            } else {
+                (ptr.offset(self.start as isize), ptr.offset(self.end as isize))
+            };
+            Drain {
+                tail:  tail,
+                start: start,
+                end:   end,
+                left:  start,
+                right: end,
+                marker: PhantomData,
+            }
+        }
+    }
+}
+
+impl DrainRange for RangeFrom<usize> {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        assert!(self.start <= vec.len(), "Range out of bounds");
+        (self.start..vec.len()).drain(vec)
+    }
+}
+
+impl DrainRange for RangeTo<usize> {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        (0..self.end).drain(vec)
+    }
+}
+
+impl DrainRange for RangeFull {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        (0..vec.len()).drain(vec)
+    }
+}
+
+impl DrainRange for usize {
+    fn drain<'a, T>(&self, vec: &'a mut Vec<T>) -> Drain<'a, T> {
+        (*self..*self+1).drain(vec)
+    }
+}
+
+/// An iterator that drains part of a vector.
 #[unsafe_no_drop_flag]
 #[unstable(feature = "collections",
            reason = "recently added as part of collections reform 2")]
 pub struct Drain<'a, T:'a> {
-    ptr: *const T,
-    end: *const T,
+    tail: usize,
+    start: *mut T,
+    end:   *mut T,
+    left:  *mut T,
+    right: *mut T,
     marker: PhantomData<&'a T>,
```

Comment:
FYI: `_x: PhantomData<T>, _y: PhantomData<&'a ()>` is not equivalent to `_xy: PhantomData<&'a T>`, because of how Sound Generic Drop treats `PhantomData<T>`.

See https://github.com/rust-lang/rfcs/blob/master/text/0769-sound-generic-drop.md#phantom-data